operations/app/src/environments/tf

```diff
@@ -158,6 +167,65 @@ else
+RESOURCE_PREFIX="$(echo "var.resource_prefix" | terraform console --var-file "${TFVARS?}" -var="terraform_caller_ip_address=${PUBLIC_IP_CIDR?}" | xargs)"
+
+# Whitelist our IP to access firewalled resources
+function add_ip_whitelist() {
+    echo -e "${GRAY?}"
+    echo -e "Whitelisting IP ${PUBLIC_IP?} for Azure resources:"
+
+    for resource in "${KeyVaults[@]}"; do
+        RESOURCE_NAME="${RESOURCE_PREFIX?}${resource?}"
+        echo -e "\t${RESOURCE_NAME?}..."
+        az keyvault network-rule add -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --ip-address "${PUBLIC_IP_CIDR?}" > /dev/null 2>&1
+    done
+
+    for resource in "${StorageAccounts[@]}"; do
+        RESOURCE_NAME="${RESOURCE_PREFIX?}${resource?}"
+        echo -e "\t${RESOURCE_NAME?}..."
+        az storage account network-rule add -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --ip-address "${PUBLIC_IP?}" > /dev/null 2>&1
+    done
+
+    for resource in "${FunctionApps[@]}"; do
+        RESOURCE_NAME="${RESOURCE_PREFIX?}${resource?}"
+        echo -e "\t${RESOURCE_NAME?}..."
+        az functionapp config access-restriction add -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --slot production --rule-name "tf-${PUBLIC_IP_HASH}" --action Allow --ip-address "${PUBLIC_IP_CIDR?}" --priority 750 > /dev/null 2>&1
+        az functionapp config access-restriction add -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --slot candidate --rule-name "tf-${PUBLIC_IP_HASH}" --action Allow --ip-address "${PUBLIC_IP_CIDR?}" --priority 750 > /dev/null 2>&1
+    done
+
+    echo -e "${NORMAL?}"
+}
+
+function remove_ip_whitelist() {
+    echo -e "${GRAY?}"
+    echo -e "Removing IP whitelist ${PUBLIC_IP?} for Azure resources:"
+
+    for resource in "${KeyVaults[@]}"; do
+        RESOURCE_NAME="${RESOURCE_PREFIX?}${resource?}"
+        echo -e "\t${RESOURCE_NAME?}..."
+        az keyvault network-rule remove -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --ip-address "${PUBLIC_IP_CIDR?}" > /dev/null 2>&1
+    done
+
+    for resource in "${StorageAccounts[@]}"; do
+        RESOURCE_NAME="${RESOURCE_PREFIX?}${resource?}"
+        echo -e "\t${RESOURCE_NAME?}..."
+        az storage account network-rule remove -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --ip-address "${PUBLIC_IP?}" > /dev/null 2>&1
+    done
+
+    for resource in "${FunctionApps[@]}"; do
+        RESOURCE_NAME="${RESOURCE_PREFIX?}${resource?}"
+        echo -e "\t${RESOURCE_NAME?}..."
+        az functionapp config access-restriction remove -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --slot production --rule-name "tf-${PUBLIC_IP_HASH}" --priority 750 > /dev/null 2>&1
+        az functionapp config access-restriction remove -g "${RESOURCE_GROUP_NAME?}" -n "${RESOURCE_NAME?}" --slot candidate --rule-name "tf-${PUBLIC_IP_HASH}" --priority 750 > /dev/null 2>&1
+    done
+
+    echo -e "${NORMAL?}"
+}
+
+trap remove_ip_whitelist EXIT
```

Comment:
Ensure that even on failure the IP addresses are removed.

As an extra safeguard, Terraform will also clear out unknown IP addresses if they persist.